exporter/datadogexporter/translate_traces.go

```diff
@@ -0,0 +1,517 @@
+
+	traces := []*pb.TracePayload{}
+
+	if resourceSpans.Len() == 0 {
+		return nil, nil
+	}
+
+	for i := 0; i < resourceSpans.Len(); i++ {
+		rs := resourceSpans.At(i)
+		if rs.IsNil() {
+			continue
+		}
+
+		// TODO: Also pass in globalTags here when we know what to do with them
+		payload, err := resourceSpansToDatadogSpans(rs, hostname, env)
+		if err != nil {
+			return traces, err
+		}
+		
+		traces = append(traces, &payload)
+		
+	}
+
+	// group the traces by env to reduce the number of flushes
+	return aggregateTracePayloadsByEnv(traces), nil
+}
+
+// converts a Trace's resource spans into a trace payload
+func resourceSpansToDatadogSpans(rs pdata.ResourceSpans, hostname string, env string) (pb.TracePayload, error) {
+	resource := rs.Resource()
+	ilss := rs.InstrumentationLibrarySpans()
+
+	// TODO: should we be getting env from the trace's spans itself? if so what's the tag+precedence order?
+	payload := pb.TracePayload{
+		HostName:     hostname,
+		Env:          env,
+		Traces:       []*pb.APITrace{},
+		Transactions: []*pb.Span{},
+	}
+
+	if resource.IsNil() && ilss.Len() == 0 {
+		return payload, nil
+	}
+
+	localServiceName, datadogTags := resourceToDatadogServiceNameAndAttributeMap(resource)
+
+	apiTraces := map[uint64]*pb.APITrace{}
+
+	for i := 0; i < ilss.Len(); i++ {
+		ils := ilss.At(i)
+		if ils.IsNil() {
+			continue
+		}
+		extractInstrumentationLibraryTags(ils.InstrumentationLibrary(), datadogTags)
+		spans := ils.Spans()
+		for j := 0; j < spans.Len(); j++ {
+			span, err := spanToDatadogSpan(spans.At(j), localServiceName, datadogTags)
+
+			if err != nil {
+				return payload, err
+			} else {
+				var apiTrace *pb.APITrace
+				var ok bool
+
+				if apiTrace, ok = apiTraces[span.TraceID]; !ok {
+					apiTrace = &pb.APITrace{
+						TraceID:   span.TraceID,
+						Spans:     []*pb.Span{},
+						StartTime: 0,
+						EndTime:   0,
+					}
+					apiTraces[apiTrace.TraceID] = apiTrace
+				}
+				
+				addToAPITrace(apiTrace, span)
+			}
+		}
+	}
+
+	for _, apiTrace := range apiTraces {
+		top := apm.GetAnalyzedSpans(apiTrace.Spans)
+		apm.ComputeSublayerMetrics(apiTrace.Spans)
+		payload.Transactions = append(payload.Transactions, top...)
+		payload.Traces = append(payload.Traces, apiTrace)
+	}
+
+	return payload, nil
+}
+
+
+// convertSpan takes an internal span representation and returns a Datadog span.
+func spanToDatadogSpan(	s pdata.Span,
+	localServiceName string,
+	datadogTags map[string]string,
+) (*pb.Span, error) {
+	tags := aggregateSpanTags(s, datadogTags)
+
+	// get tracestate as just a general tag
+	if len(s.TraceState()) > 0 {
+		tags[tracetranslator.TagW3CTraceState] = string(s.TraceState())
+	}
+
+	// get start/end time to calc duration
+	startTime := s.StartTime()
+	endTime := s.EndTime()
+	duration := int64(endTime) - int64(startTime)
+
+	// it's possiblle end time is unset, so default to 0 rather than using a negative number
+	if s.EndTime() == 0 {
+		duration = int64(0)
```

Comment:
I think we can do
```suggestion
		duration = 0
```
can't we?